### PR TITLE
FIX BUG(merge?) J=ceil(log(N)/log(2))

### LINKED ISSUE #363: Scattering2D doesn't work with 2**J == image_size
I'm not sure if #346 fixes #284 :

```
import torch
from kymatio import Scattering2D

scattering = Scattering2D(J=5, shape=(32, 32))
x = torch.randn(1, 1, 32, 32)
Sx = scattering(x)
print(Sx.size())
```
gives:
```
C:\Python36\python.exe D:/Cours/3A/PFE/Python/kymatio_examples/scattering2D_test.py
Traceback (most recent call last):
  File "D:/Cours/3A/PFE/Python/kymatio_examples/scattering2D_test.py", line 6, in <module>
    Sx = scattering(x)
  File "C:\Python36\lib\site-packages\kymatio-0.2.0.dev0-py3.6.egg\kymatio\scattering2d\scattering2d.py", line 235, in __call__
  File "C:\Python36\lib\site-packages\kymatio-0.2.0.dev0-py3.6.egg\kymatio\scattering2d\scattering2d.py", line 188, in forward
  File "C:\Python36\lib\site-packages\kymatio-0.2.0.dev0-py3.6.egg\kymatio\scattering2d\backend\backend_torch.py", line 40, in __call__
  File "C:\Python36\lib\site-packages\torch\nn\modules\module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "C:\Python36\lib\site-packages\torch\nn\modules\padding.py", line 172, in forward
    return F.pad(input, self.padding, 'reflect')
  File "C:\Python36\lib\site-packages\torch\nn\functional.py", line 2685, in pad
    ret = torch._C._nn.reflection_pad2d(input, pad)
RuntimeError: Argument #4: Padding size should be less than the corresponding input dimension, but got: padding (32, 32) at dimension 3 of input [1, 1, 32, 32]

Process finished with exit code 1
```
Can you reproduce the error ?
The goal is to get a (1, K, 1, 1) tensor.

Thanks

### PR DESCRIPTION
**Pull request by edouardoyallon**
Close #363

### CODE REVIEW

File: kymatio/scattering2d/scattering2d.py [hunk 1]
```diff
@@ -92,6 +92,8 @@ def __init__(self, J, shape, L=8, max_order=2, pre_pad=False):
         self.pre_pad = pre_pad
         self.max_order = max_order
         self.shape = shape
+        if 2**J>shape[0] or 2**J>shape[1]:
+            raise (RuntimeError('The largest dimension should be smaller than 2^J'))
```
Comment (Jonas1312) [inline kymatio/scattering2d/scattering2d.py]: should be greater than

Comment (edouardoyallon) [reply kymatio/scattering2d/scattering2d.py]: you mean like "2**J>=shape[0] or 2**J>=shape[1]" ?! Isn't it precisely the setting you're trying to use(and that was available with prior scatnet software)?

Comment (edouardoyallon) [reply kymatio/scattering2d/scattering2d.py]: my bad

**Comment by Jonas1312** [comment]
Right now:
```
import torch
from kymatio import Scattering2D

scattering = Scattering2D(J=5, shape=(32, 32))
x = torch.randn(1, 1, 32, 32)
Sx = scattering(x)
print(Sx.size())
a = Sx.view(-1)
print(a.size())
```

gives the following output:
```
torch.Size([1, 1, 681, 0, 0])
torch.Size([])
```

I think the proper output size should be:
```
torch.Size([1, 1, 681, 1, 1])
```

**Comment by edouardoyallon** [comment]
@Jonas1312 thanks a lot, I would have missed it.. this is due to the unpad function. So here, I see 3 options:

1. Maintaining a padding by 3*N=3*2^J as initially done.
2. Modifying the unpad function such that it handles non-symmetric padding. The issue that comes to my mind is that it will lead to an assymetry which is not really desirable (it was actually a bug previously introduced in scatnet-2D ..)
3. Editing special routines for the the setting where N=2^J. Indeed, in this setting, the aliasing will be potentially strong and that could be an issue... For instance, using a stronger oversampling and a simple sum as a low-pass filtering could be an improvement.

@janden @lostanlen @eickenberg  any suggestions? This should strategy could be common to 1-2-3-D signals..

**Comment by eickenberg** [comment]
So while looking at this, I went to verify that we have the option of doing the padding ourselves. We do seem to have that, but in a weird way in which the fact of "not padding" is given to the padding object which then doesn't pad :)

https://github.com/kymatio/kymatio/blob/master/kymatio/scattering2d/backend/backend_torch.py#L41

While we're at it, the casting to complex performed in the padding object seems completely out of scope for that function. Also `unpad` seems surprisingly simple as a function - I'm surprised unpadding is always achieved by just removing a pixel everywhere, but if it works, cool!

I have no clear opinion about this issue as of now, because I don't really know all the things that had been done up to now, but I do think that

- signal sizes of size `2**J` will probably happen a lot, especially for researchers and people from signal processing that want to control exactly what is going on, so this case should be treated with care
- cases that are not of size `2**J` shouldn't be treated radically differently without reason/explanation from the former. If suddenly the padding doubles, it's weird and completely unpredictable for a user

**Comment by edouardoyallon** [comment]
okay, for now I'll go to solution 1. I'll try to reproduce the @joanbruna 's experiments with textures

**Comment by lostanlen** [comment]
i'm a bit late to the party, but why does this have to be a special case? 
can't we have `J=Inf` for expected scattering and `J` finite for everything that requires spatial localization?

**Comment by edouardoyallon** [comment]
@lostanlen the main reason is that pytorch knows how to pad a signal of size N up to... 2N-1, but it launches an error on 2N ... it's not a difficult fix but it has to be done carefully.. this is why it's still not done yet.

**Comment by edouardoyallon** [comment]
hi @eickenberg is it mergeable? if yes, please merge.

**Comment by eickenberg** [comment]
if travis passes, we can merge this

**Pull request merged**